internals/overlord/logstate/loki_test.go

```diff
@@ -0,0 +1,175 @@
+		Service: "svc1",
+		Message: "log line #5",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 55, 0, time.UTC),
+		Service: "svc3",
+		Message: "log line #6",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 56, 0, time.UTC),
+		Service: "svc2",
+		Message: "log line #7",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 57, 0, time.UTC),
+		Service: "svc1",
+		Message: "log line #8",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 58, 0, time.UTC),
+		Service: "svc4",
+		Message: "log line #9",
+	}}
+
+	expected := `
+{
+  "streams": [
+    {
+      "stream": {
+        "pebble_service": "svc1"
+      },
+      "values": [
+          [ "1704026090000000000", "log line #1" ],
+          [ "1704026092000000000", "log line #3" ],
+          [ "1704026094000000000", "log line #5" ],
+          [ "1704026097000000000", "log line #8" ]
+      ]
+    },
+    {
+      "stream": {
+        "pebble_service": "svc2"
+      },
+      "values": [
+          [ "1704026091000000000", "log line #2" ],
+          [ "1704026096000000000", "log line #7" ]
+      ]
+    },
+    {
+      "stream": {
+        "pebble_service": "svc3"
+      },
+      "values": [
+          [ "1704026093000000000", "log line #4" ],
+          [ "1704026095000000000", "log line #6" ]
+      ]
+    },
+    {
+      "stream": {
+        "pebble_service": "svc4"
+      },
+      "values": [
+          [ "1704026098000000000", "log line #9" ]
+      ]
+    }
+  ]
+}
+`
+
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		c.Assert(r.Method, Equals, http.MethodPost)
+		c.Assert(r.Header.Get("Content-Type"), Equals, "application/json")
+
+		reqBody, err := io.ReadAll(r.Body)
+		c.Assert(err, IsNil)
+		expFlattened, err := flattenJSON(expected)
+		c.Assert(err, IsNil)
+		c.Assert(string(reqBody), Equals, expFlattened)
+	}))
+	defer s.Close()
+
+	cl := newLokiClient(&plan.LogTarget{Location: s.URL})
+	for _, entry := range input {
+		err := cl.Write(context.Background(), entry)
+		c.Assert(err, IsNil)
+	}
+
+	err := cl.Flush(context.Background())
+	c.Assert(err, IsNil)
+}
+
+func (*lokiSuite) TestLokiFlushCancelContext(c *C) {
+	serverCtx, killServer := context.WithCancel(context.Background())
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		select {
+		case <-serverCtx.Done():
+		// Simulate a slow-responding server
+		case <-time.After(10 * time.Second):
+		}
+	}))
+	defer s.Close()
+	defer killServer()
+
+	cl := newLokiClient(&plan.LogTarget{Location: s.URL})
+	err := cl.Write(context.Background(), servicelog.Entry{
+		Time:    time.Now(),
+		Service: "svc1",
+		Message: "this is a log line\n",
+	})
+	c.Assert(err, IsNil)
+
+	// Cancel the Flush context quickly
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Microsecond)
+	defer cancel()
+
+	flushReturned := make(chan struct{})
+	go func() {
+		err = cl.Flush(ctx)
+		close(flushReturned)
```

Comment:
I think this close should be after the `c.Assert` below, otherwise I think there's a race: the `c` instance might be done / cleaned up before we assert on it. I think we've had issues with this kind of thing in Juju tests with goroutines.

Reply:
ah, subtle